.github/workflows/checks.yml

```diff
@@ -3,27 +3,282 @@ name: Checks
 
 jobs:
-  test:
+  linux_tests:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
+      - name: Configure environment variables
+        run: |
+          # echo "::set-env name=SHELLFLAGS::-x"
+          echo "::set-env name=BUILDFLAGS::"
+          echo "::set-env name=RSPY_TARGET_DIR::${GITHUB_WORKSPACE}/target"
+          echo "::set-env name=CARGO_TARGET_DIR::${GITHUB_WORKSPACE}/target"
+      # Necessary for now for the cargo cache:
+      # https://github.com/actions/cache/issues/133#issuecomment-599102035
+      - name: Fix ~/.cache permissions
+        run: |
+          sudo chown -R $(whoami):$(id -ng) ~/.cargo/
+      - name: Cache cargo registry
+        uses: actions/cache@v1
+        with:
+          path: ~/.cargo/registry
+          key: ${{ runner.os }}-cargo-registry-${{ hashFiles('**/Cargo.toml') }}-5-
+      - name: Cache cargo index
+        uses: actions/cache@v1
+        with:
+          path: ~/.cargo/git
+          key: ${{ runner.os }}-cargo-index-${{ hashFiles('**/Cargo.toml') }}-5-
+      - name: Cache cargo target
+        uses: actions/cache@v1
+        with:
+          path: ${{ github.workspace }}/target
+          key: ${{ runner.os }}-cargo-target-${{ hashFiles('**/Cargo.toml') }}-5-
+      - name: Cache cargo rslib
+        uses: actions/cache@v1
+        with:
+          path: ${{ github.workspace }}/rslib/target
+          key: ${{ runner.os }}-cargo-rslib-${{ hashFiles('**/Cargo.toml') }}-5-
+      - name: Cache cargo rspy
+        uses: actions/cache@v1
+        with:
+          path: ${{ github.workspace }}/rspy/target
+          key: ${{ runner.os }}-cargo-rspy-${{ hashFiles('**/Cargo.toml') }}-5-
+      - name: Cache pip wheels
+        uses: actions/cache@v1
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}-5-
+      - name: Cache pyenv
+        uses: actions/cache@v1
+        with:
+          path: ${{ github.workspace }}/pyenv
+          key: ${{ runner.os }}-pyenv-${{ hashFiles('**/Makefile') }}-${{ hashFiles('**/requirements.txt') }}-5-
       - name: Set up python
         uses: actions/setup-python@v1
         with:
           python-version: 3.7
       - name: Set up protoc
-        uses: ankitects/setup-protoc@master
+        uses: arduino/setup-protoc@v1.1.0
         with:
           repo-token: ${{ secrets.GITHUB_TOKEN }}
       - name: Set up node
         uses: actions/setup-node@v1
         with:
           node-version: 12
-      - name: Run checks
+      - name: Set up ripgrep, pyaudio, gettext, rename
         run: |
-          # add requirements
-          sudo apt update; sudo apt install portaudio19-dev gettext rename
+          sudo apt update
+          sudo apt install portaudio19-dev gettext rename
           sudo snap install ripgrep --classic
-          export CARGO_TARGET_DIR=~/target
-          export RSPY_TARGET_DIR=~/target
-          make check build BUILDFLAGS=""
```

Comment:
The build has gone from 10 minutes to 16 because you're doing an optimized build now. I'm not sure it's worth paying that price just to have the artefacts available for every PR? It may make more sense to split the artefact generation up into a separate workflow that only runs on tagged commits, and runs for both Python 3.7 and 3.8.

Reply:
(though the caching you've added looks to have improved the macOS build considerably - well done!)

Reply:
Very impressed with the speed improvements you've made - I don't think there's any need to split out the artefact generation now.

Reply:
From the tests I made with the last two commits, the caches are not fully working for Mac OS. Here are the results:
1. The `make develop` on Mac OS, Linux and Windows used the cached results 100%. Their build times where respectively 1:19, 2:01 and 2:19 minutes. Where Mac OS is the fasted from them all.
1. The `make build` is super simple and just takes about 50 seconds in all of them.
1. The `make check` is troublesome on Mac OS. On Windows and Linux, the Rust Libraries caches worked and their build times were 2:34 and 2:12 minutes. But on Mac OS, the Rust Libraries caches were ignored and the build time took 5:50 minutes.

I will try to change the make rules to `make check`, `make develop` and `make build` and see if the Mac OS caches start working. If I remember correctly, the problem of doing `make check develop build`, is that `make check` causes `make develop` and `make build` to take 0 seconds, i.e., `make check` builds everything requires by those two other rules.

Reply:
I just confirmed after testings:
1. https://github.com/ankitects/anki/pull/528/checks?check_run_id=535187485#step:18:20
1. https://github.com/ankitects/anki/runs/534904775?check_suite_focus=true#step:18:22

That what was causing Mac OS Rust libraries cache to not be used was using the `RSPY_TARGET_DIR` and `CARGO_TARGET_DIR`. While using `CARGO_TARGET_DIR` and `RSPY_TARGET_DIR` are working fine for Linux and Windows, on Mac OS they are causing the cache to not be used, i.e., to be rebuilt on the next step which tries to use it.

As a simple fix, I just remove the `RSPY_TARGET_DIR` and `CARGO_TARGET_DIR` from the Mac OS matrix build. This is not a problem because Mac OS is the fastest of them all and can build the caches into `rslib/target` and `rspy/target` without problems. And directories `rspy/target` and `rslib/target` are also already cached, so, Mac OS will only take a little longer to build when the cache is reset. Most the time the Mac OS builds are going to be using the caches on `rslib/target` and `rspy/target`.

As of now, the times to fully build everything are: https://github.com/ankitects/anki/pull/528/checks?check_run_id=535187459
1. 7:48 minutes on Mac OS (with all caches)
1. 9:05 minutes on Linux (with all caches)
1. 10:52 minutes on Windows (with all caches)

When all caches are reset (a full build without any cache), the times are: https://github.com/ankitects/anki/runs/534849168
1. 18:04 minutes on Mac OS (build all caches)
1. 24:22 minutes on Linux (build all caches)
1. 26:43 minutes on Windows (build all caches)